h2/utilities.py

```diff
@@ -215,6 +225,24 @@ def _reject_uppercase_header_fields(headers, hdr_validation_flags):
         yield header
 
 
+def _reject_surrounding_whitespace(headers, hdr_validation_flags):
+    """
+    Raises a ProtocolError if any header name or value is surrounded by
+    whitespace characters.
+    """
+    for header in headers:
+        if header[0] and ((header[0][0] in _WHITESPACE) or
+                          (header[0][-1] in _WHITESPACE)):
+            raise ProtocolError(
+                "Received header name surrounded by whitespace %r" % header[0])
+        if header[1] and ((header[1][0] in _WHITESPACE) or
```

Comment:
Hrm. RFC 7230 allows empty header field values (though I wish it didn't), so we have to allow those too. However, we don't have to allow empty header field names, so let's remove that check and then, if that test still fails, we can amend the Hypothesis testing strategy to forbid empty header field names.